src/main/java/com/github/maxopoly/finale/misc/VelocityHandler.java

```diff
@@ -1,33 +1,70 @@
 package com.github.maxopoly.finale.misc;
 
-import java.util.Collection;
 import java.util.Map;
 import java.util.Set;
-import java.util.TreeSet;
 
+import org.bukkit.craftbukkit.v1_13_R2.entity.CraftPlayer;
+import org.bukkit.craftbukkit.v1_13_R2.entity.CraftProjectile;
 import org.bukkit.entity.EntityType;
 import org.bukkit.entity.Player;
 import org.bukkit.entity.Projectile;
 
+import net.minecraft.server.v1_13_R2.EntityPlayer;
+import net.minecraft.server.v1_13_R2.EntityPotion;
+import net.minecraft.server.v1_13_R2.EntityProjectile;
+import net.minecraft.server.v1_13_R2.MathHelper;
+
 public class VelocityHandler {
 
 	private Set<EntityType> typesToRevert;
+	private Map<EntityType, Double> horizontalMultiplier;
+	private Map<EntityType, Double> verticalMultiplier;
+	private Map<EntityType, Float> pitchOffsets;
+	private Map<EntityType, Float> powers;
+	
 	private Map<EntityType, Double> velocityMultiplier;
-
-	public VelocityHandler(Collection<EntityType> revertedTypes, Map<EntityType, Double> velocityMultiplier) {
-		this.typesToRevert = new TreeSet<>();
+	
+	public VelocityHandler(Set<EntityType> revertedTypes, Map<EntityType, Double> velocityMultiplier, Map<EntityType, Double> horizontalMultiplier, Map<EntityType, Double> verticalMultiplier,
+			Map<EntityType, Float> pitchOffsets, Map<EntityType, Float> powers) {
+		this.typesToRevert = revertedTypes;
 		this.velocityMultiplier = velocityMultiplier;
+		this.horizontalMultiplier = horizontalMultiplier;
+		this.verticalMultiplier = verticalMultiplier;
+		this.pitchOffsets = pitchOffsets;
+		this.powers = powers;
 	}
 
 	public void modifyVelocity(Projectile projectile, Player shooter) {
-		if (typesToRevert.contains(projectile.getType())) {
-			projectile.setVelocity(projectile.getVelocity().subtract(shooter.getVelocity()));
+		if (typesToRevert.contains(projectile.getType()) && powers.containsKey(projectile.getType())) {
+			float power = powers.get(projectile.getType());
+			float pitchOffset = pitchOffsets.containsKey(projectile.getType()) ? pitchOffsets.get(projectile.getType()) : 1;
+			double horizontal = horizontalMultiplier.containsKey(projectile.getType()) ? horizontalMultiplier.get(projectile.getType()) : 1;
+			double vertical = verticalMultiplier.containsKey(projectile.getType()) ? verticalMultiplier.get(projectile.getType()) : 1;
+			shoot(projectile, shooter, power, pitchOffset, horizontal, vertical);
 		}
 		Double multiplier = velocityMultiplier.get(projectile.getType());
 		if (multiplier == null) {
 			return;
 		}
 		projectile.setVelocity(projectile.getVelocity().multiply(multiplier));
 	}
+	
+	public void shoot(Projectile projectile, Player pShooter, float power, float pitchOffset, double horizontal, double vertical) {
```

Comment:
This method deserves a lot of comments

Reply:
Copied projectile shooting NMS method from 1.7

Reply:
At least note that then, any future maintainer finding this method is gonna curse you if there's not at least some indication on what the heck this is.